larsim/EventGenerator/POTaccumulator_module.cc

```diff
@@ -0,0 +1,307 @@
+
+
+  /// Prints the list of subruns with partial or missing POT information.
+  void printMissingSubrunList() const;
+
+  /// Prints the list of POT per run.
+  void printRunSummary() const;
+
+  /// Prints the total POT summary.
+  void printSummary() const;
+
+
+  /// Converts the information from `POT` in a compact string.
+  static std::string to_string(sumdata::POTSummary const& POT);
+
+}; // class sim::POTaccumulator
+
+
+//------------------------------------------------------------------------------
+//---  module implementation
+//---
+//------------------------------------------------------------------------------
+sim::POTaccumulator::POTaccumulator(Parameters const& config)
+  : EDAnalyzer(config)
+  , fPOTtag(config().SummaryTag())
+  , fSummaryOutputCategory(config().SummaryCategory())
+  , fRunOutputCategory(config().RunSummaryCategory())
+  {}
+
+
+//------------------------------------------------------------------------------
+void sim::POTaccumulator::beginSubRun(art::SubRun const& subRun) {
+
+  auto const& ID = subRun.id();
+
+  //
+  // get the information from the subrun and update the subrun counts
+  //
+  art::Handle<sumdata::POTSummary> summaryHandle;
+  if (!subRun.getByLabel(fPOTtag, summaryHandle)) {
+    ++fMissingSubrunFragments[ID];
+    mf::LogDebug(fSummaryOutputCategory) << "Fragment of subrun " << ID
+      << " has no '" << fPOTtag.encode() << "' POT summary.";
+    return;
+  }
+
+  ++fPresentSubrunFragments[ID];
+
+  //
+  // accumulate the information by run
+  //
+  auto const& runID = subRun.getRun().id();
+
+  sumdata::POTSummary const& subRunPOT = *summaryHandle;
+
+  fRunPOT[runID].aggregate(subRunPOT);
+  MF_LOG_TRACE(fSummaryOutputCategory) << "Fragment #"
+    << fPresentSubrunFragments[ID] << " of subrun " << ID
+    << ": " << sim::POTaccumulator::to_string(subRunPOT)
+    ;
+
+} // sim::POTaccumulator::analyze()
+
+
+//------------------------------------------------------------------------------
+void sim::POTaccumulator::endJob() {
+
+  //
+  // print the run summary
+  //
+
+  if (!fRunOutputCategory.empty()) {
+
+    if (!fMissingSubrunFragments.empty()) printMissingSubrunList();
+
+    printRunSummary();
+
+  } // if run report requested
+
+
+  //
+  // print the total summary
+  //
+  printSummary();
+
+} // sim::POTaccumulator::endJob()
+
+
+//------------------------------------------------------------------------------
+void sim::POTaccumulator::printMissingSubrunList() const {
+
+  //
+  // missing fragments information
+  //
+  mf::LogVerbatim log{fRunOutputCategory};
+  log << size(fMissingSubrunFragments) << " subruns miss POT information:";
+
+  auto const efound = fPresentSubrunFragments.cend();
+
+  for (auto const& [ id, nMissing ]: fMissingSubrunFragments) {
+
+    // add to the count of fragments the ones which did have
+    unsigned int nFragments = nMissing;
+    auto const iFound = fPresentSubrunFragments.find(id);
+    if (iFound != efound) nFragments += iFound->second;
```

Comment:
This logic doesn't make sense to me.  Under what circumstance do you expect a `SubRunID` that does *and* does not contain a POT summary object?  And if such a situation can occur, why report the total number of fragments instead of just the number that are actually missing?

Reply:
In case one input file with a fragment of that subrun (from one input file) contains such information, and another (from a different input file) does not.
I think there is some value in seeing at once that a subrun is _partially_ missing information.
To be honest, I think that may be an error where files from different samples are mixed, with overlapping subrun numbers.
This all section of missing information is a debugging tool: IMO this information should not be missing. But I don't feel that strong enough to throw exceptions.